### PR TITLE
sql: disallow DESC option in the last column of inverted indexes

### LINKED ISSUE #84388: sql: descending inverted indexes do not work
An example:

```
defaultdb> CREATE TABLE t (k INT PRIMARY KEY, j JSON, INVERTED INDEX j_idx (j DESC));
CREATE TABLE

defaultdb> INSERT INTO t VALUES (1, '[0, 1, 2, 3]'), (2, '[0, 2, 4]'), (3, '[1, 2, 4]');
INSERT 3

defaultdb> SELECT * FROM t@j_idx WHERE j @> '2';
ERROR: error decoding 6 bytes: did not find marker 0x12 in buffer 0x3700012a0400
```

I'm not sure, but it seems like we never intended to allow the DESC option for inverted indexes. I don't see any use of specifying the direction of an inverted index, because we cannot perform a range scan on an inverted index. So I think we should disallow DESC (and maybe ASC too to prevent confusion) for inverted index columns. I don't think we need to worry about this being a breaking change, since these indexes are broken already.

What about multi-column inverted indexes? Should we allow the non-inverted columns to be DESC? Again, range scans on inverted indexes are not possible, so I don't think there is any use-case for this. However, I don't think it's broken so maybe we should just allow it? If we disallow it, would we need to consider a migration strategy because we consider it backward incompatible?

Jira issue: CRDB-17631

### PR DESCRIPTION
**Pull request by mgartner**
The last column of inverted indexes cannot have the `DESC` direction.
Prior to this commit it was allowed, but caused internal errors. To
avoid propagating the notion that inverted indexes have a semantic
direction, we stop printing the `ASC` option for inverted index columns
in the output of `SHOW CREATE TABLE`.

Fixes #84388

Release note (sql change): The last column of an INVERTED INDEX can no
longer have the `DESC` option. If `DESC` was used in prior versions, it
could cause internal errors.

### CODE REVIEW

**Review by rytaft — COMMENTED** [review]
> Reviewed 16 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84516)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @mgartner, and @michae2)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84516#-N7332xq2Z0w8lxwWyPn):*
I don't think this is true. There are some cases where we scan a range. For example, look at the second span here:
```
> EXPLAIN (opt) SELECT a FROM json_tab WHERE b @> '{"a": {}}' ORDER BY a
;
                                 info
----------------------------------------------------------------------
  sort
   └── project
        └── inverted-filter
             ├── inverted expression: /5
             │    ├── tight: true, unique: false
             │    └── union spans
             │         ├── ["7a\x00\x019", "7a\x00\x019"]
             │         └── ["7a\x00\x02\x00\xff", "7a\x00\x03")
             └── scan json_tab@foo_inv
                  └── inverted constraint: /5/1
                       └── spans
                            ├── ["7a\x00\x019", "7a\x00\x019"]
                            └── ["7a\x00\x02\x00\xff", "7a\x00\x03")
```
Why can't we just disallow `DESC` for now?

**Comment by mgartner** [comment]
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84516#-N7332xq2Z0w8lxwWyPn):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I don't think this is true. There are some cases where we scan a range. For example, look at the second span here:
```
> EXPLAIN (opt) SELECT a FROM json_tab WHERE b @> '{"a": {}}' ORDER BY a
;
                                 info
----------------------------------------------------------------------
  sort
   └── project
        └── inverted-filter
             ├── inverted expression: /5
             │    ├── tight: true, unique: false
             │    └── union spans
             │         ├── ["7a\x00\x019", "7a\x00\x019"]
             │         └── ["7a\x00\x02\x00\xff", "7a\x00\x03")
             └── scan json_tab@foo_inv
                  └── inverted constraint: /5/1
                       └── spans
                            ├── ["7a\x00\x019", "7a\x00\x019"]
                            └── ["7a\x00\x02\x00\xff", "7a\x00\x03")
```
Why can't we just disallow `DESC` for now?
</blockquote></details>

Good point. I still think it's odd to give the impression that inverted index columns have a semantic ordering. The indexed column contains only one dimension of the value, and an inverted index could never support sort-free ordering on the inverted index column.

Postgres disallows ASC on GIN indexes, which is one of the reasons I removed it:

```
marcus=# create table t (a int[]);
CREATE TABLE

marcus=# create index on t using gin (a asc);
ERROR:  0A000: access method "gin" does not support ASC/DESC options
LOCATION:  ComputeIndexAttrs, indexcmds.c:2033
```

But allowing it would be more backward compatible...

@cockroachdb/sql-experience any thoughts?

**Comment by rafiss** [comment]
assorted thoughts:
- Are we able to use any form of product telemetry to determine how common `using gin (a asc)` expressions are?
- What if we showed a NOTICE saying that ASC is ignored (if it is specified), rather than an error?
- When it comes to backward compat concerns, let's involve the Product team. cc @vy-ton

**Comment by mgartner** [comment]
I've updated this PR to only disallow `DESC` and stop printing `ASC` in the output of `SHOW CREATE TABLE` for inverted index columns. This should not be considered a breaking change because creating an inverted index column with `DESC` would cause internal errors prior to this commit.

**Review by rytaft — APPROVED** [review]
:lgtm:

> Reviewed 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84516#-:-N7QljO41RFbb13ItShC:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @michae2)

File: pkg/sql/logictest/testdata/logic_test/inverted_index [hunk 1]
```diff
@@ -1689,5 +1707,5 @@ SELECT primes FROM cb WHERE primes && numbers ORDER BY primes
 
 # Regression test for incorrectly unwrapping a DOidWrapper (#84569).
 statement ok
-CREATE TABLE t84569 (name_col NAME NOT NULL, INVERTED INDEX (name_col gin_trgm_ops DESC));
```
Comment (mgartner) [inline pkg/sql/logictest/testdata/logic_test/inverted_index:1692]: I've confirmed that the bug was reproducible without `DESC`, so this regression test remains useful.

**Comment by mgartner** [comment]
TFTR!

bors r+

**Pull request merged**